File: typescript/src/calisthenics-exercises/first-class-collections.ts
export type Product = { id: string; price: number }

export class ProductOperations {
  private products: Product[] = []

  addProduct(product: Product): void {
    const exists = this.products.some((p) => p.id === product.id)
    if (!exists) this.products.push(product)
  }

  totalPrice(): number {
    return this.products.map((p) => p.price).reduce((a, b) => a + b, 0)
  }

  removeProduct(productId: string): void {
    this.products = this.products.filter((p) => p.id !== productId)
  }
}

File: csharp/src/calisthenics-exercises/FirstClassCollections.cs
namespace RefactoringAvanzado.CalisthenicExercises;

public class Product
{
    public string Id { get; set; }
    public double Price { get; set; }

    public Product(string id, double price)
    {
        Id = id;
        Price = price;
    }
}

public class ProductOperations
{
    private readonly List<Product> _products = new();

    public void AddProduct(Product product)
    {
        var exists = _products.Any(p => p.Id == product.Id);
        if (!exists) _products.Add(product);
    }

    public double TotalPrice()
    {
        return _products.Sum(p => p.Price);
    }

    public void RemoveProduct(string productId)
    {
        _products.RemoveAll(p => p.Id == productId);
    }
}
